### DISCUSSION TITLE
HTTP

### DISCUSSION THREAD

**Discussion by jangwooooo** [subak-ba/cs-study#56] [네트워크(응용 계층)]
## HTTP
HTTP는 HyperText Transfer Protocol의 약자이며, W3 상에서 정보를 주고받을 수 있는 프로토콜이다.
주로 TCP를 사용하고 80번 포트를 사용한다.  

HTTP는 클라이언트와 서버 사이에 이루어지는 요청/응답(request/response) 프로토콜입니다.
클라이언트가 HTTP를 통하여 서버로부터 정보를 요청하면, 서버는 이 요청에 응답하여 필요한 정보를 해당 사용자에게 전달하게 됩니다.  

HTTP를 통해 전달되는 자료는 http:로 시작하는 URL로 조회할 수 있습니다.

<br>

## HTTP Header
사용자가 HTTP 요청을 하게 되면 header와 body를 주고 받습니다.  

body는 본문이라고 할 수 있으며 서버에서 보내고자 하는 컨텐츠 본문인 JSON, html, image등이 담깁니다.
header는 body를 설명하는 정보를 포함해서 여러가지 정보가 담긴 정보묶음입니다.

헤더는 콜론 ':' 으로 서로 구분되는 key - value형태로 설정됩니다.
HTTP요청을 할 때 3가지의 헤더인 일반헤더, 요청헤더, 응답헤더가 자동으로 생깁니다.
여기서 서버에서 설정하는 헤더를 응답헤더, 클라이언트에서 설정한 헤더를 요청헤더라고 합니다.  

HTTP 헤더자체가 굉장히 유연하게 설계가 되어있어서 커스텀하게 만들 수 있습니다.
하지만 보통은 지정되어있는 key값에 value를 담아서 헤더를 설정합니다.

### 일반 헤더
<img width="793" alt="image" src="https://github.com/subak-ba/cs-study/assets/108380536/7b22f905-dc41-4ce2-a23a-302d1e872d4e">  

### 요청 헤더
<img width="1128" alt="image" src="https://github.com/subak-ba/cs-study/assets/108380536/8fd10400-c7cc-4e07-91f1-a0b1a5e5eb4c">  

### 응답 헤더
<img width="998" alt="image" src="https://github.com/subak-ba/cs-study/assets/108380536/5ee402e7-b054-4977-987e-69297a4cf69e">  

<br>
<br>
<br>

## HTTP/1.0과 HTTP/1.1의 차이
HTTP/1.0의 문제를 보완한 것이 HTTP/1.1입니다.
두 버전의 차이를 알아보겠습니다.  

### HTTP/1.0
HTTP/1.0은 연결의 수명이 짧습니다.
각 HTTP 요청당 TCP 핸드셰이크가 발생되며 기본적으로 한 연결당 하나의 요청을 처리하도록 설계되었습니다.  
한 번 연결할 때마다 TCP연결을 계속해야 하니 RTT가 늘어나는 문제점이 있습니다.  
이를 해결하기 위해 HTTP/1.1이 나옵니다.

<img width="783" alt="image" src="https://github.com/subak-ba/cs-study/assets/108380536/ee9dc72b-97e5-4d39-b4d5-67447787295d">   

<br>

### HTTP/1.1
HTTP/1.1은 HTTP/1.0의 단점을 보완한 프로토콜입니다.  
크게 3가지의 차이점이 있습니다.  

1. keep-alive default
매번 데이터를 요청할 때마다 TCP 연결을 하는게 아닌 한 번 연결을 해놓고 계속해서 데이터를 받을 수 있게 만들었습니다.  
이는 keep-alive 옵션을 기본옵션으로 하면서 가능해졌습니다.  
keep-alive header를 통해 TCP 연결유지시간인 timeout과 최대 요청수 max를 정할 수 있습니다.  

2. 호스트 헤더
HTTP/1.0은 서버가 하나의 호스트만 가진다고 가정하기 때문에 HTTP/1.0은 헤더에 호스트를 포함하지 않았습니다. 
이 때문에 HTTP/1.0은 하나의 IP에 하나의 호스트만 가질 수 있었습니다.
그러나 사실 서버는 여러개의 호스트를 가질 수 있으며 이런 유연성을 위해 HTTP/1.1은 헤더에 특정 호스트를 포함할 수 있게 변경되었습니다. 
HTTP/1.1에서는 항상 호스트를 포함해서 요청하도록 바뀌었습니다.  

3. 대역폭 최적화
HTTP/1.0의 경우 어떠한 파일을 다운로드 받다가 연결이 끊기면 이어서 다시 다운로드 받는 것은 불가능했습니다. 
하지만 HTTP/1.1에서는 연결해서 받는 게 가능하도록 만들었습니다. 
예를 들어 HTTP/1.0에서는 10KB 파일을 다운로드 받는다고 했을 때 5KB까지만 받고 다시 다운로드를 받는게 불가능했습니다.  
이를 HTTP/1.1에서는 Range:bytes=5000- 라는 헤더를 추가해서 다운로드 재개 요청을 할 수 있게 바뀐 것입니다.    

### 요청을 줄이기 위한 기술
HTTP/1.1로 발전했음에도 불구하고 서버요청할 때마다 RTT는 계속해서 증가하기 때문에 요청을 줄이기 위한 여러가지 기술들이 있습니다.  
대표적으로 이미지스프라이트, 코드압축, Base64 인코딩 기술을 같이 쓰게 되었습니다. 

### HTTP/1.1의 고질적인 문제 : HOL
HTTP/1.1에서는 HOL과 무거운 헤더(Header)를 가지는 문제점이 있었고 이를 해결하지 못했습니다.
HOL(Head Of Line Blocking)이란 네트워크에서 같은 큐에 있는 패킷이 그 첫번째 패킷에 의해 지연될 때 발생하는성능저하현상을 말합니다.  

<img width="633" alt="image" src="https://github.com/subak-ba/cs-study/assets/108380536/9f87dac9-2996-4ba1-9bd1-154ce3c393d9">  

<br>
<br>
<br>

### HTTP/2
2009년 구글은 HTTP/1.1의 한계를 극복하기 위해 SPDY 프로토콜을 개발했습니다.  
이 후 2015년, SPDY를 기반으로 하는 HTTP/2 프로토콜을 만들었습니다.  

HTTP/2는 HTTP/1.1의 단점을 보완한 프로토콜입니다.   

### 바이너리 포맷 계층
<img width="854" alt="image" src="https://github.com/subak-ba/cs-study/assets/108380536/fdad4d90-4952-4624-b95a-d919356b6a0b">  
  
### 멀티플렉싱
단일 TCP연결의 여러 스트림에서 여러 HTTP 요청과 응답을 비동기적으로 보낼 수 있습니다.  
이를 통해 HOL을 해결합니다.

HTTP1.1에서는 병렬요청을 하려면 다중TCP연결을 통해서 하고 일반적으로는 TCP연결하나당 병렬요청은 불가능했습니다.

이를 HTTP/2.0에서는 리소스를 작은 프레임으로 나누고 이를 스트림으로 프레임을 전달합니다. 
각각의 프레임은 스트림ID, 해당 청크의 크기를 나타내는 프레임이 추가되었기 때문에 작게 나눠서 다운로드가 되더라도 결과적으로 응답데이터에서는 올바른 순서로 재조립할 수 있게 됩니다.  

<img width="831" alt="image" src="https://github.com/subak-ba/cs-study/assets/108380536/9fcda7a4-7be0-4f18-8d45-11db9b61955b">. 

### 서버 푸쉬
<img width="625" alt="image" src="https://github.com/subak-ba/cs-study/assets/108380536/64e7c36e-46bf-4ffe-bd4b-46fd17062159">
  
### 헤더 압축
HTTP/1.1에서는 무거운 헤더가 있었지만 HTTP/2에서는 이를 허프만 인코딩 압축 방법으로 압축시킵니다.  
똑같은 서버에서 2개의 이미지를 준다고 했을 때 중복되는 헤더는 제외한채 보내고 해당 공통 필드로 헤더를 재구성하며 중복되지 않은 헤더값은 허프만 인코딩 압축 방법으로 압축해 전송합니다.  
 
<br>

## HTTP/3 
HTTP/2는 여전히 TCP를 사용하기 때문에 초기 연결에 대한 RTT로 인한 지연시간이라는 문제점이 있었습니다. 
HTTP/3은 이를 해결한 버전입니다.  

QUIC(Quick UDP Internet Connections)이라는 계층 위에서 돌아가며, TCP 기반이 아닌
UDP 기반으로 돌아 가며 HTTP/2에서 장점이었던 멀티플렉싱 등을 가지고 있으며
초기연결설정시 지연시간 감소라는 대표적 특성을 가지고 있습니다.  

또한, 전송된 패킷이 손실 되었다면 수신측에서 에러를 검출하고 수정하는 방식이며 열악한 네트워크 환경에서도 낮은 패킷손실률을 자랑하는 순방향 오류 수정 메커니즘(FEC, Forward Error Correction)이라는 특징을 가집니다.

**Comment by BLoHny** [comment]
> HTTP/2은 이를 해결한 버전입니다.

오타가 있습니다.

**Reply by jangwooooo** [reply]
감사합니다. 수정하였습니다.
